### PR TITLE
Add Dodo to dex.trades in Dune v2 for Ethereum Mainnet.

### PR DESCRIPTION
**Pull request by LeviScoffie**
Brief comments on the purpose of your changes:

* I added the Dodo dex trades on Ethereum mainnet to the dex.trades table on V2.
* Defined all my sources and described the columns from resultant files in my schema file.
* Altered the dbt.projects yml file to include the dodo_ethreum new table.
* Updated the dex.trades sql file with the new Dodo trades.
* Run `dbt compile` and got feedback message `done`
.

*For Dune Engine V2*
I've checked that:
General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributors)
* [x] if adding a new materialized table, I edited the optimize table macro

Join logic:
* [x] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [x] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [x] where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

File: macros/optimize_tables.sql [hunk 1]
```diff
@@ -167,6 +167,9 @@ OPTIMIZE sudoswap_ethereum.pool_trades;
 OPTIMIZE airswap_ethereum.trades;
 {% endset %}
 
+{% set dodo_ethereum_trades %}
+OPTIMIZE dodo_ethereum_trades;
```
Comment (Hosuke) [inline macros/optimize_tables.sql:171]: Here may be `dodo_ethereum.trades`, but this file has been changed in the latest version of main branch.

File: models/dodo/ethereum/dodo_ethereum_trades.sql [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        when lower(token_a_symbol) > lower(token_b_symbol) then concat(token_b_symbol, '-', token_a_symbol)
+        else concat(token_a_symbol, '-', token_b_symbol)
+    end AS token_pair,
+    token_a_amount AS token_bought_amount,
+    token_b_amount AS token_sold_amount,
+    token_a_amount_raw AS token_bought_amount_raw,
+    token_b_amount_raw AS token_sold_amount_raw,
+    usd_amount AS amount_usd,
+    token_a_address AS token_bought_address,
+    token_a_address AS token_sold_address,
```
Comment (Hosuke) [inline models/dodo/ethereum/dodo_ethereum_trades.sql:338]: Here I made a mistake too.
It should be `token_b_address AS token_sold_address,`

**Review by Hosuke — COMMENTED** [review]
I found out few typo but overall impressive.

File: models/dodo/ethereum/dodo_ethereum_schema.yml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      - &token_bought_amount_raw
+        name: token_bought_amount_raw
+        description: "Raw value of the token bought at time of execution without division by its decimals"
+      - &token_sold_amount_raw
+        name: token_sold_amount_raw
+        description: "Raw value of the token sold at time of execution without division by its decimals"
+      - &amount_usd
+        name: amount_usd
+        description: "USD value of the trade at time of execution"
+      - &token_bought_address
```
Comment (jeff-dude) [inline models/dodo/ethereum/dodo_ethereum_schema.yml:68]: please apply the generic test for token_bought on this field, example:
https://github.com/duneanalytics/spellbook/blob/main/models/uniswap/ethereum/uniswap_ethereum_schema.yml#L64

in order for it to run, values need manually added to the seed file:
https://github.com/duneanalytics/spellbook/blob/main/seeds/dex/trades/dex_trades_seed.csv

most dexes have analytics sites that link tx's from the dex to etherscan, where you can grab the necessary values to populate the seed. you can also query the source decoded table for that tx_hash to find the values to put in place.

the goal is to ensure we aren't flipping token bought/sold on the fields in the model

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: models/dodo/ethereum/dodo_ethereum_trades.sql (hunk 2 above)
Comment (jeff-dude) [reply models/dodo/ethereum/dodo_ethereum_trades.sql:338]: good catch -- exactly why we apply the generic test with values in the seed!

File: models/dodo/ethereum/dodo_ethereum_trades.sql [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        ) as usd_amount,
+        token_a_address,
+        token_b_address,
+        exchange_contract_address,
+        tx_hash,
+        tx.`from` as tx_from,
+        tx.`to` as tx_to,
+        trace_address,
+        evt_index,
+        row_number() OVER (PARTITION BY project, tx_hash, evt_index, trace_address ORDER BY version, category) AS trade_id
```
Comment (jeff-dude) [inline models/dodo/ethereum/dodo_ethereum_trades.sql:69]: this column is no longer necessary. please remove from select statements, unique keys in config, schema file and anywhere else applicable

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: models/dodo/ethereum/dodo_ethereum_trades.sql (hunk 2 above)
Comment (LeviScoffie) [reply models/dodo/ethereum/dodo_ethereum_trades.sql]: Thanks  alot @Hosuke 

@jeff-dude  forgive my lack of proactivity. I was out just got back home. Let me do the changes.

File: models/dodo/ethereum/dodo_ethereum_trades.sql (hunk 4 above)
Comment (jeff-dude) [reply models/dodo/ethereum/dodo_ethereum_trades.sql]: nvm, i got this covered in latest commit

Comment (LeviScoffie) [reply models/dodo/ethereum/dodo_ethereum_trades.sql]: Okay. Let me do that asap.

**Comment by jeff-dude** [comment]
@LeviScoffie please see my latest commit to standardize the script. the issue now is duplicates produced in the output. please see the attached dbt slim ci job (can click details).

here is query to help:
```
select
  block_date,
  blockchain,
  project,
  version,
  tx_hash,
  evt_index,
  trace_address,
  count(1)
from
  dbt_jeff_dodo_ethereum.trades
group by
  block_date,
  blockchain,
  project,
  version,
  tx_hash,
  evt_index,
  trace_address
having
  count(*) > 1
```

not sure where the rows get duplicated, please do some research and let me know

File: models/dodo/ethereum/dodo_ethereum_schema.yml (hunk 3 above)
Comment (LeviScoffie) [reply models/dodo/ethereum/dodo_ethereum_schema.yml:67]: Gotcha

**Comment by LeviScoffie** [comment]
Yeah sure let me try figure it out.

**Comment by LeviScoffie** [comment]
Should we define the incremental strategy to overwrite instead of appending?

**Comment by LeviScoffie** [comment]
@jeff-dude I'm parsing the script in dune but the query time is running out. What should I do to ensure I get results withing query runtime?

**Comment by Hosuke** [comment]
> 



> @LeviScoffie please see my latest commit to standardize the script. the issue now is duplicates produced in the output. please see the attached dbt slim ci job (can click details).
> 
> here is query to help:
> 
> ```
> select
>   block_date,
>   blockchain,
>   project,
>   version,
>   tx_hash,
>   evt_index,
>   trace_address,
>   count(1)
> from
>   dbt_jeff_dodo_ethereum.trades
> group by
>   block_date,
>   blockchain,
>   project,
>   version,
>   tx_hash,
>   evt_index,
>   trace_address
> having
>   count(*) > 1
> ```
> 
> not sure where the rows get duplicated, please do some research and let me know

In my code for airswap trades I still use trade_id as part of unique key. So the records in airswap_ethereum.trades with trade_id > 1 are the duplicated records in original decoded table. I have filtered them in this query https://dune.com/queries/1376296 .
I am not sure if this is the similar situation as in airswap trades.
If I need to remove the trade_id column and duplicated records I will update in the next PR.

**Comment by LeviScoffie** [comment]
@Hosuke but I think the trace ID field is null all over the dex.trades tables? Or what i'm I missing?

**Comment by Hosuke** [comment]
> @Hosuke but I think the trace ID field is null all over the dex.trades tables? Or what i'm I missing?
`trace_id` is null, while `trade_id` is the ranked id I use to distinguish the duplicated records in the original table.

**Comment by LeviScoffie** [comment]
@Hosuke  sorry my bad. The trade_id field is used to rank the duplicate trades events with similar transaction details based on the partitions in  the window function .I've added it to the tests in the schema file. Ty you for pointing me to the right direction pal. We could make a PR to delete the records later based on @jeff-dude 's guidance.

**Comment by jeff-dude** [comment]
> > 
> 
> > @LeviScoffie please see my latest commit to standardize the script. the issue now is duplicates produced in the output. please see the attached dbt slim ci job (can click details).
> > here is query to help:
> > ```
> > select
> >   block_date,
> >   blockchain,
> >   project,
> >   version,
> >   tx_hash,
> >   evt_index,
> >   trace_address,
> >   count(1)
> > from
> >   dbt_jeff_dodo_ethereum.trades
> > group by
> >   block_date,
> >   blockchain,
> >   project,
> >   version,
> >   tx_hash,
> >   evt_index,
> >   trace_address
> > having
> >   count(*) > 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > not sure where the rows get duplicated, please do some research and let me know
> 
> In my code for airswap trades I still use trade_id as part of unique key. So the records in airswap_ethereum.trades with trade_id > 1 are the duplicated records in original decoded table. I have filtered them in this query https://dune.com/queries/1376296 . I am not sure if this is the similar situation as in airswap trades. If I need to remove the trade_id column and duplicated records I will update in the next PR.

i'd like to dig into that further. why are the rows duplicated in the original decoded table(s)? this seems to apply both to airswap & dodo here. while adding a ```trade_id``` in the unique key values helps avoid failures in data ingestion, it then produces duplicates in dex trades. if others build a dashboard on dex trades, but don't know to filter out rows where ```trade_id > 1```, we will have inflated metrics

i would like to see an opposite approach taken, where we identify duplicates in the model and filter out prior to loading the tables. does this mean we need more columns from the decoded table(s) in the unique key? are the rows expected to be duplicated, if so, why?

**Comment by jeff-dude** [comment]
i would like to eliminate the ```trade_id``` approach altogether, as it doesn't help solve our needs. we could use similar logic to help filter duplicates, but don't apply in table downstream, if necessary

**Comment by jeff-dude** [comment]
> 



> > i would like to eliminate the `trade_id` approach altogether, as it doesn't help solve our needs. we could use similar logic to help filter duplicates, but don't apply in table downstream, if necessary
> 
> Yes. The same logic could be used to filter for only the rows with `trade_id`=1 then not include them in schema.yml and final table.

this would be our last resort ideally. can you answer why the duplicates exist in the first place? is that expected and i'm just unaware of how the data flows into those decoded tables? or are we able to tweak the decoded table queries to fix duplicates at the source?

**Comment by jeff-dude** [comment]
> @jeff-dude @LeviScoffie For airswap, the code `row_number() OVER (PARTITION BY tx_hash, evt_index, trace_address) AS trade_id` exists in original deprecated v1 script
> 
> https://github.com/duneanalytics/spellbook/blob/49964675327d46400255272fd11dffe7f79365b3/deprecated-dune-v1-abstractions/ethereum/dex/trades/insert_airswap.sql#L56
> 
> 
> When I first tried migrating the script to v2, I eliminated the trade_id, and then I fail with the [unique key test](https://github.com/duneanalytics/spellbook/pull/1669/checks?check_run_id=8621952224)
> So then I realized there are duplicated rows in original decoded tables, but not sure why, then I keep the `trade_id` as part of unique key.
> In [v1 docs](https://docs.dune.com/data-tables/abstractions/dex.trades) this column is described as 'for database magic'.
> If we can fix duplicates at the source then will be perfect to remove `trade_id` column.
> Maybe I need to study some more about decoding process.

yes, we'd want to fix at decoding if there are truly duplicates. the intention of ```trade_id``` in v1 engine wasn't for unique or not (there were unique indexes on the table instead).

i will open new PR for airswap discussion

as for dodo (@LeviScoffie) there are still dupes, i can add context shortly

**Comment by jeff-dude** [comment]
@LeviScoffie here is another query to view duplicates:
https://dune.com/queries/1383165?d=1

notice the fields in group by for ```dupes``` CTE, these are the fields we need unique across all dexes in dex trades

it looks like one of the v2 decoded tables is causing dupes, please continue analysis from here and let me know if it's a bug with decoded table(s) or something else

**Comment by LeviScoffie** [comment]
> > @jeff-dude @LeviScoffie For airswap, the code `row_number() OVER (PARTITION BY tx_hash, evt_index, trace_address) AS trade_id` exists in original deprecated v1 script
> > 
> > https://github.com/duneanalytics/spellbook/blob/49964675327d46400255272fd11dffe7f79365b3/deprecated-dune-v1-abstractions/ethereum/dex/trades/insert_airswap.sql#L56
> > 
> > 
> > When I first tried migrating the script to v2, I eliminated the trade_id, and then I fail with the [unique key test](https://github.com/duneanalytics/spellbook/pull/1669/checks?check_run_id=8621952224)
> > So then I realized there are duplicated rows in original decoded tables, but not sure why, then I keep the `trade_id` as part of unique key.
> > In [v1 docs](https://docs.dune.com/data-tables/abstractions/dex.trades) this column is described as 'for database magic'.
> > If we can fix duplicates at the source then will be perfect to remove `trade_id` column.
> > Maybe I need to study some more about decoding process.
> 
> yes, we'd want to fix at decoding if there are truly duplicates. the intention of ```trade_id``` in v1 engine wasn't for unique or not (there were unique indexes on the table instead).
> 
> i will open new PR for airswap discussion
> 
> as for dodo (@LeviScoffie) there are still dupes, i can add context shortly

Please add  the duplicates for more clarity. I think the duplicates originate from the source of the decoded tables.

**Comment by LeviScoffie** [comment]
> @LeviScoffie here is another query to view duplicates: https://dune.com/queries/1383165?d=1
> 
> notice the fields in group by for `dupes` CTE, these are the fields we need unique across all dexes in dex trades
> 
> it looks like one of the v2 decoded tables is causing dupes, please continue analysis from here and let me know if it's a bug with decoded table(s) or something else

Apparently there are some decoded tables  for DODO in v2 that have dupes and others do not. I tried this query 
```select
  block_date,
  blockchain,
  project,
  version,
  tx_hash,
  evt_index,
  trace_address,
  count(1)
from
  dbt_jeff_dodo_ethereum.trades
group by
  block_date,
  blockchain,
  project,
  version,
  tx_hash,
  evt_index,
  trace_address
having
  count(*) > 1```  
for one of the tables and the duplicate count was 33 for each entry. https://dune.com/queries/1373000

**Comment by jeff-dude** [comment]
thank you. i believe i spotted the fix. i'll push a commit here shortly. it's not actually an issue with the decoded table(s)

**Comment by LeviScoffie** [comment]
> thank you. i believe i spotted the fix. i'll push a commit here shortly. it's not actually an issue with the decoded table(s)

Great. Happy to collab. What's the actual issue?

**Comment by jeff-dude** [comment]
i take it back, there appears to be two issues. i pushed one fix on the prices join (needed filter for 'ETH', otherwise returned all 33 assets where blockchain was null).

second issue are these three decoded tables, they appear to overlap tx's:
https://dune.com/queries/1383165

we need to investigate if those three are decoded correctly, and if so, then do we need ```union``` vs ```union all```? why do the tx's overlap?

**Comment by LeviScoffie** [comment]
Let me find out from the contract's source code and the protocol's docs and discord. Could this be a case of smart contract routing?

**Comment by LeviScoffie** [comment]
I have requested for help in the decode channel. We could get further clearance on the issue from there.

**Comment by LeviScoffie** [comment]
Is it a problem with filtering for the ":blockchain" contract address just as you did in the airswap model @Hosuke @jeff-dude

**Comment by Hosuke** [comment]
> Is it a problem with filtering for the ":blockchain" contract address just as you did in the airswap model @Hosuke @jeff-dude

I check the [dupes in DODO](https://dune.com/queries/1387409), the token addresses appear mainly only on ethereum.
@LeviScoffie you can cross check the token address in [this view](https://dune.com/queries/1385967)
I think it's different situation as in airswap.

**Comment by jeff-dude** [comment]
> Is it a problem with filtering for the ":blockchain" contract address just as you did in the airswap model @Hosuke @jeff-dude

i added the additional filters for blockchain on the erc20 token view joins, but this is not the issue. you can see from linked query above that dupes happen prior to this join

**Comment by jeff-dude** [comment]
looks like the legend @antonio-mendes has found the issue in the decoding discord channel for you -- thanks both for digging into it!

**Comment by LeviScoffie** [comment]
Yes I've been chatting with @antonio-mendes and we agreed that I would resubmit the contracts again for decoding. Apprentlly the same contract addresses shared names. Like DSP and DPP both having a similar address. I also did further digging and found out that most of the submitted contracts were minimal proxies of base contracts. So i've prepared a csv with the main smart contracts that provide the base logic for the minimal proxies. Waiting on his advice on the use of minimal proxies and base logic contracts.

**Comment by LeviScoffie** [comment]
> Yes I've been chatting with @antonio-mendes and we agreed that I would resubmit the contracts again for decoding. Apprentlly the same contract addresses shared names. Like DSP and DPP both having a similar address. I also did further digging and found out that most of the submitted contracts were minimal proxies of base contracts. So i've prepared a csv with the main smart contracts that provide the base logic for the minimal proxies. Waiting on his advice on the use of minimal proxies and base logic contracts.

See this [query](https://dune.com/queries/1389570)

**Comment by LeviScoffie** [comment]
Hi @jeff-dude  the DVM,DSP,DPP contracts I have discovered they route through the proxy contracts emitting Order history events. I think dropping them would greatly help in curbing this issue.

**Comment by antonio-mendes** [comment]
I think me and @LeviScoffie figured out how to get these contracts decoded properly on Dune. I'm working on it today.
Will let you know here when it's live 🤗

**Comment by jeff-dude** [comment]
thank you both 🙌 feel free to tag me when resolved, i think we'll be close to merge

**Comment by LeviScoffie** [comment]
> thank you both raised_hands feel free to tag me when resolved, i think we'll be close to merge

I was trying to see a change in use of `union` vs `union all`.

**Comment by jeff-dude** [comment]
@LeviScoffie looks like the team fixed source data issues 🚀 one final thing to add, based on this comment:
https://github.com/duneanalytics/spellbook/pull/1691#discussion_r990404903

a few entries into our seed file, values grabbed from third party source, to ensure token bought is aligned correctly. once added to file, you can apply test in your schema file, as noted in comment above.

**Comment by LeviScoffie** [comment]
Great..thats good news.. Right on it ser. Incase of any snarl ups I''ll  lyk.

**Review by jeff-dude — APPROVED** [review]

**Comment by antonio-mendes** [comment]
Thanks again for the patience with this @LeviScoffie ! Super happy to see this live 🎉

**Comment by LeviScoffie** [comment]
> Thanks again for the patience with this @LeviScoffie ! Super happy to see this live 🎉 

No thank you, @antonio-mendes for the immense support you and the team offered. We are live!

**Pull request merged**